crowbar_framework/lib/repository.rb

```diff
@@ -0,0 +1,144 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Repository
+  class Check
+    attr_reader :platform, :id, :name, :required, :available, :md5, :url
+
+    def initialize(platform, repo)
+      @platform = platform
+      @id = repo
+      @name = config_repo["name"]
+      @required = config_requirement
+      @available = validate_repo_dir
+      @md5 = validate_key_file
+      @url = config_web_url
+    end
+
+    protected
+
+    #
+    # config
+    #
+    def config
+      YAML.load_file(Repository.config_path)
+    end
+
+    def config_repo
+      config[@platform]["repos"][@id]
+    end
+
+    def config_repomd_md5
+      config_repo["repomd"]["md5"]
+    end
+
+    def config_repomd_key
+      config_repo["repomd"]["key"] || "/srv/tftpboot/#{@platform}/repos/#{@name}/repodata/repomd.xml.key"
+    end
+
+    def config_requirement
+      config_repo["required"]
+    end
+
+    def config_web_url
+      config_repo["url"] || \
+        "http://#{NodeObject.admin_node.ip}:#{Repository.web_port}/#{@platform}/repos/#{@name}"
+    end
+
+    #
+    # validation helpers
+    #
+    def all_repos_dir
+      path = File.join("/srv/tftpboot", @platform, "repos")
+      Dir["#{path}/*"].map { |p| p.split("/").last }
+    end
+
+    def key_file
+      if File.exist?(config_repomd_key)
+        config_repomd_key
+      else
+        false
+      end
+    end
+
+    #
+    # validations
+    #
+    def validate_key_file
+      key = key_file
+      if key
+        md5 = Digest::MD5.hexdigest(File.read(key))
+        config_repomd_md5.include? md5
+      else
+        false
+      end
+    end
+
+    def validate_repo_dir
+      all_repos_dir.include?(@name)
+    end
+  end
+
+  extend self
+
+  #
+  # generic repo methods
+  #
+  def config_path
+    File.join(Rails.root, "config", "repos.yml")
+  end
+
+  def check_all_repos
+    repochecks = []
+    all_platforms.each do |platform|
+      repositories(platform).each do |repo|
+        check = Repository::Check.new(platform, repo)
+        repochecks << check
+      end
+    end
+    repochecks
+  end
+
+  def where(options = {})
+    platform = options.fetch :platform, nil
+    repo = options.fetch :repo, nil
+    if platform
```

Comment:
I think you should move these `if` statements inside the `check_all_repos.select` loop so you don't have to duplicate it.